### ISSUE TITLE
[#470] [review-finding] P100: llama-server router mode never starts Hydra RPC listener — migration/P-D decode to P100 fails (SocketException 111)

### ISSUE THREAD

**Issue by ddvnguyen** [ddvnguyen/hydra_vortex#577]
## Finding

P100 migration and P/D decode are BROKEN on the current deployment: the coordinator cannot open the Hydra RPC connection to the P100 node. Root cause: fork #90 swapped the P100 image from `llama-engine` to `llama-server`, and `tools/server/server.cpp` starts the Hydra RPC listener ONLY in the non-router (`else`) branch. P100 runs in ROUTER MODE (no boot model; preset-p100.ini on-demand load), so `--rpc-port 9502` is accepted but never acted on. The `llama-engine` binary starts RPC unconditionally (`llama-engine.cpp:836`), which is why RTX (9503) works and P100 (9502) does not.

## Evidence (live rig, 2026-08-07)

- `SessionMigrationMidWorkflow` solo on a clean rig: Phase 1 (5 turns on rtx) PASSES; `POST /sessions/{id}/migrate {"target":"p100"}` → HTTP 500, empty body, 261ms.
- Core log: `System.Net.Sockets.SocketException (111): Connection refused` in `MigrateSessionAsync` (WorkerSchedulerService.cs:302) — coordinator → 192.168.122.21:9502 refused.
- Host probes: `192.168.122.21:8086` (HTTP) OPEN; `192.168.122.21:9502` and `127.0.0.1:9502` CLOSED. RTX `127.0.0.1:9503` OPEN.
- VM (192.168.122.21): `ss -tlnp` shows only 8086 (llama HTTP), 9700 (head), exporters — NO 9502.
- P100 llama-server startup (journalctl): `starting router server, no model will be loaded` → `router server is listening on http://0.0.0.0:8086`. NO `hydra rpc: unified server on 0.0.0.0:9502` line. Binary `build 9668 (f6dd4e52c)` — note: image tag `sm60-llama-server-6d00536` but built from `f6dd4e52c` (PR #91), one merge behind the epic head `6d0053691` (build 9670 on RTX).

## Root cause

`src/llama-cpp/tools/server/server.cpp:269-289` — router branch starts HTTP only; `if (params.rpc_port > 0) ctx_server.start_rpc_server(...)` at :336-337 lives ONLY in the `else` (non-router) branch. `start_rpc_server` itself has no mode gating and logs on both success/failure — the absence of the 'unified server' line proves it was never invoked.

## Impact

- `SessionMigrationMidWorkflow` (and any rtx→P100 migration / P/D decode via RPC) fails 100%.
- P100 still answers `/health` / `/slots` (HTTP) so the coordinator reports it `healthy` — a healthy-but-RPC-dead node.

## Suggested fix (fork; needs lead decision — options)

1. Move/duplicate the `start_rpc_server` call into the router branch in `server.cpp` (one line; matches llama-engine behavior). OR
2. Revert P100 to the `llama-engine` image (fork #90 revert). OR
3. Give P100 a boot model (non-router mode) — heavier.

Also flag: image `sm60-llama-server-6d00536` was built from `f6dd4e52c` but tagged as the head — verify the build pipeline tags by actual commit, not intended head.

---
**Part of epic #470** ([Llama-Engine] Merge/validate P/D STATE_PUT + decode RPC)

**Comment by ddvnguyen** [comment]
Fixed + verified live on the epic branch (commit 33ce8a0b4).

**Fix:** P100 now runs **llama-engine** (not llama-server):
- `infra/hydra-head/config/node-p100.yaml` → image `sm60-llama-engine-6d00536`, binary/dest `llama-engine`, plus boot-resident `model: /mnt/readonly_data/Qwopus3.6-35B-A3B-v1-APEX-I-Balanced.gguf` (has_model mode; Core skips inline reload on p100). Kept `models-preset` for alias resolution.
- `.github/workflows/deploy-heads.yml` → p100 image is the engine twin (remap fork's post-#90 server output); dropped the server-forcing override.
- `scripts/deploy-hydra-head.sh` + `scripts/ci/list-llama-image-tags.sh` + ci.yml comments updated; PROJECT_STATUS.md verified-fact row added.

**Verification (live, 2026-08-07):**
- P100 head restarted → pulled `sm60-llama-engine-6d00536` (digest sha256:150cb131…) → `build 9670 (6d0053691)` running.
- `hydra rpc: unified server on 0.0.0.0:9502` + accept loop started; host `192.168.122.21:9502` now OPEN (was connection refused).
- Model loaded: `Qwopus3.6-35B-A3B-v1-APEX-I-Balanced.gguf`; /v1/models lists it.
- Coordinator /health: p100 healthy, 1 slot idle.
- P/D decode on P100 works end-to-end: coordinator routed a session `prefill_node=rtx decode_node=p100` with `restore_kv_ms=198`, 100 tokens out, non-empty reply (auto P/D migration at turn 2).

**Note on deploy pitfall:** the head skips the OCI pull when `dest` already exists and no digest/checksum is configured — a stale llama-engine binary (build 9668, pre-#92) was silently reused on first restart. Deleted it to force the re-pull. Recommend adding a checksum/digest to node-p100.yaml so future image pin changes always re-pull.

**Fork follow-up (not in this commit):** revert #90 in `hydra-build.yml` so the fork's `p100-image` output is natively the engine image; deploy-heads.yml currently remaps it.

**Comment by ddvnguyen** [comment]
**Lead update (2026-08-08):** Live-rig analysis shows the deployed engines are build 9670 = fork commit 6d0053691, which does NOT include llama.cpp PR #93 (DECODE_APPLY restore check inversion + trailing-logits) or issue #94 gaps (all OPEN in fork). The p100 merged-decode HTTP 400 on `GET /v1/decode/{id}` is suspected to be the DECODE_APPLY restore path on the sm60 engine. Path forward: fork PR #93 (gap impl in progress) → merge → submodule bump → deploy → re-test. Baseline LiveRig run 31242336898 dispatched to capture current state.

**Comment by ddvnguyen** [comment]
**Root cause CONFIRMED (2026-08-08, live p100 engine log):**

```
hydra: DECODE_APPLY slot=0 v2 blob: restored 20 tokens from header   ← restore OK
hydra: DECODE_APPLY slot=0 KV restore failed (66082684)              ← inverted check reports failure
done request: GET /v1/decode/13482 ... 400                           ← 400 to coordinator
```

The sm60 p100 engine (build 9670 = fork commit 6d0053691) **successfully restores the KV blob, but its DECODE_APPLY restore check is inverted / compares against the wrong length** — the exact bug fixed by llama.cpp PR #93 (commits 63666cd42 + 0a0fa9a72) and hardened by issue #94 gaps. Deployed engines predate those commits.

**Fix path:** PR #93 (now with the 3 gap fixes, head 4dcd1f2e8) → review in progress → merge to hydra-fork → bump submodule on epic/470-merged-decode → rebuild + deploy engines → re-run LiveRig. This is in motion.
